### ISSUE TITLE
Language introspection

### ISSUE THREAD

**Issue by robrix** [maxbrunsfeld/tree-sitter#3]
Nodes have a count and a named count, but it’d be useful to be able to tell this a priori about the production for a given `TSSymbol` (if I’m understanding these pieces correctly).

**Comment by maxbrunsfeld** [comment]
Yeah, this makes sense. Maybe these functions?

```c
size_t ts_language_symbol_count(const TSLanguage *);
const char * ts_language_symbol_name(const TSLanguage *, TSSymbol);
bool ts_language_symbol_is_terminal(const TSLanguage *, TSSymbol);
```

The last one would require adding a new field to the `TSLanguage` struct, and changing the code generator to populate that field. They all should be pretty easy to add though.

One concern about these functions is their names potentially conflict with the names of the exported language functions like `ts_language_javascript()`. Maybe it would make sense to change the exported names to something like `ts_javascript_language()`.

**Comment by robrix** [comment]
:+1: to the suggested functions.

> One concern about these functions is their names potentially conflict with the names of the exported language functions like `ts_language_javascript()`. Maybe it would make sense to change the exported names to something like `ts_javascript_language()`.

That could work. Alternatively you could provide a way to override the exported symbol in case of collision.

**Comment by robrix** [comment]
Two out of three of those APIs are now in place :white_check_mark:

**Comment by robrix** [comment]
@maxbrunsfeld: Any objections to me tackling the `ts_language_is_terminal` one?

**Comment by maxbrunsfeld** [comment]
Yeah, I'd love it if you took that on! Let me know if you want to screenshare or chat about it at any point.

**Comment by robrix** [comment]
:bow: will do!

**Comment by maxbrunsfeld** [comment]
A couple of thoughts on this:

Currently, the only data that is provided to the runtime about each symbol is an enum value called [TSNodeType](https://github.com/maxbrunsfeld/tree-sitter/blob/master/include/tree_sitter/parser.h#L23):
  * `Named` (regular keys in the grammar's `rules` map)
  * `Anonymous` (inline strings or regexes in the rules)
  * `Hidden` (underscore-prefixed keys in the `rules`, or auxiliary rules generated by the compiler).

In order to add a new independent field to the metadata we store about each symbol, it probably makes sense to replace this enum with a bitfield of boolean flags, like this:

```c
typedef struct {
  
  // Is this symbol exposed via the TSNode API?
  // * `false` for symbols with underscore-prefixed rule names in the grammar
  // * `false` for auxiliary symbols generated by the compiler
  // * `true` otherwise
  bool visible: 1;

  // Is this symbol exposed by the `named` family of TSNode APIs?
  // * `true` for symbols corresponding to rule names in the grammar
  // * `false` for symbols corresponding to inline strings and regexps
  bool named: 1;

  // Will nodes with this symbol ever have children?
  // * `true` for symbols that end up in the `SyntaxGrammar`
  // * `false` for symbols that end up in the `LexicalGrammar`
  bool terminal: 1;

} TSSymbolMetadata;
```

Currently, if a symbols is not 'visible', we don't care whether it's named or not, but conceptually they're independent, so this representation seems a bit clearer.

**Comment by maxbrunsfeld** [comment]
For the change in question though, I don't think it's necessary to actually track whether `hidden` symbols are `named` or not; that would requires a few more changes in the compiler that aren't really related to our main goal here. You could just set `named` to `false` if `visible` is false.

**Comment by robrix** [comment]
Thanks for the details! That’s super helpful :sparkles:

**Comment by maxbrunsfeld** [comment]
I think the idea of a `terminal` bit has been superseded by the `node-types.json` file, which provides general metadata about the shape of each node type.
